introduce the hubspot engagement table to adjust the joins in int_hub…

fivetran-reneeli opened the pull request:
## PR Overview
**This PR will address the following Issue/Feature:** #11  #10 #12 

**This PR will result in the following new package version:**
<!--- Please add details around your decision for breaking vs non-breaking version upgrade. If this is a breaking change, were backwards-compatible options explored? -->
v0.1.0-a4

**Please provide the finalized CHANGELOG entry which details the relevant changes included in this PR:**
<!--- Copy/paste the CHANGELOG for this version below. -->


## Breaking Changes
- Added the hubspot `engagement` table to the package and made the following updates:
    - Added `stg_rag_hubspot__engagement` model as part of the hubspot staging models.
    - Updated `int_rag_hubspot__deal_document` to adjust the method that `hubspot_engagement_*` models are joined by leveraging the `hubspot__engagement` table as the intermediary joining table for the `engagement_contact` and `engagement_company` tables.
    - Updated `int_rag_hubspot__deal_document` to retrieve `engagement_type` from the hubspot `engagement` table as opposed to the `engagement_emails` and `engagement_notes` tables. As such, removes their respective references as they are no longer used in this model.

## Under the Hood
- Updated the unique key in `rag__unified_document` to include `chunk_index`. Previously, the unique key was a combination of only `document_id`, `platform`, and `source_relation`, which was potentially inaccurate if there were multiple chunks associated with a document.

## PR Checklist
### Basic Validation
Please acknowledge that you have successfully performed the following commands locally:
- [x] dbt run –full-refresh && dbt test
- [na] dbt run (if incremental models are present) && dbt test

Before marking this PR as "ready for review" the following have been applied:
- [x] The appropriate issue has been linked, tagged, and properly assigned
- [x] All necessary documentation and version upgrades have been applied
- [x] docs were regenerated (unless this PR does not include any code or yml updates)
- [x] BuildKite integration tests are passing
- [x] Detailed validation steps have been provided below

### Detailed Validation
Please share any and all of your validation steps:
<!--- Provide the steps you took to validate your changes below. -->


### If you had to summarize this PR in an emoji, which would it be?
<!--- For a complete list of markdown compatible emojis check our this git repo (https://gist.github.com/rxaviers/7360908)  --> 
:dancer:

fivetran-reneeli commented on models/intermediate/hubspot/int_rag_hubspot__deal_document.sql:
Actually I realized that now that `engagement` exists and the field `engagement_type` is available in that table, do we still need the coalesce between `"engagement_emails.engagement_type", "engagement_notes.engagement_type"` here https://github.com/fivetran/dbt_unified_rag/blob/96524a7a6da9f0bc669692a3e5edfa509084fd3e/models/intermediate/hubspot/int_rag_hubspot__deal_document.sql#L56

in order to grab `engagement_type`

fivetran-reneeli replied:
update-- have since removed the previous `"engagement_emails.engagement_type", "engagement_notes.engagement_type"` to swap with `engagement.engagement_type`

fivetran-reneeli commented on CHANGELOG.md:
```diff
@@ -1,3 +1,14 @@
+    - Updated `int_rag_hubspot__deal_document` to adjust the method that `hubspot_engagement_*` models are joined by leveraging the `hubspot__engagement` table as the intermediary joining table for the `engagement_contact` and `engagement_company` tables.
+    - Updated `int_rag_hubspot__deal_document` to retrieve `engagement_type` from the hubspot `engagement` table as opposed to the `engagement_emails` and `engagement_notes` tables. As such, removes their respective references as they are no longer used in this model.
+
+## Under the Hood
+- Updated the unique key in `rag__unified_document` to include `chunk_index`. Previously, the unique key was a combination of only `document_id`, `platform`, and `source_relation`, which was potentially inaccurate if there were multiple chunks associated with a document.
```
Good point, updated

fivetran-reneeli commented on dbt_project.yml:
```diff
@@ -37,5 +38,5 @@ models:
       +materialized: view
     staging:
-      +materialized: view
+      +materialized: table
```
oops leftover -- removed

fivetran-reneeli commented on models/intermediate/hubspot/int_rag_hubspot__deal_document.sql:
```diff
@@ -64,24 +69,21 @@ engagement_detail_prep as (
     left join companies
         on engagement_companies.company_id = companies.company_id
         and engagement_companies.source_relation = companies.source_relation
-    left join engagement_emails
-        on engagement_deals.engagement_id = engagement_emails.engagement_id
-        and engagement_deals.source_relation = engagement_emails.source_relation
-    left join engagement_notes
-        on engagement_deals.engagement_id = engagement_notes.engagement_id
-        and engagement_deals.source_relation = engagement_notes.source_relation
```
ah good catch! removed

fivetran-reneeli commented on dbt_project.yml:
```diff
@@ -2,7 +2,7 @@ config-version: 2
 
 name: 'unified_rag'
-version: '0.1.0'
+version: '0.1.0-a4'
```
Ohh my bad. I'll update that everywhere except in the release tag

fivetran-avinash commented on macros/staging/hubspot/get_hubspot_engagement_columns.sql:
```diff
@@ -0,0 +1,16 @@
+{% macro get_hubspot_engagement_columns() %}
+
+{% set columns = [
```
I see some columns in here (`owner_id`, `active`, `created_at`) that aren't present in the seed or the staging model, should we remove them here?

fivetran-avinash commented on integration_tests/seeds/hubspot_engagement.csv:
```diff
@@ -0,0 +1,2 @@
+id,type,_fivetran_synced,portal_id
```
Is there a reason only a subset of seed columns are being run and tested here? Can we add all the relevant columns brought into the staging model?

fivetran-reneeli replied:
that's a great point.  i initially mocked this up from the internal schema with the new version of tables that i was testing with. I added in the rest of the columns!

fivetran-avinash commented on models/staging/hubspot_staging/stg_rag_hubspot.yml:
```diff
@@ -96,6 +96,39 @@ models:
+          The ID of the engagement's owner.
+
+          PLEASE NOTE - This field will only be populated for pre HubSpot v3 API versions. This field is only included to allow for backwards compatibility between HubSpot API versions. This field will be deprecated in the near future.
+      - name: portal_id
```
Don't forget to document `source_relation`!

fivetran-reneeli replied:
thanks! updated

fivetran-avinash commented on CHANGELOG.md:
```diff
@@ -1,3 +1,14 @@
+## Breaking Changes
+- Added the hubspot `engagement` source table to the package and made the following updates:
+    - Added `stg_rag_hubspot__engagement` model as part of the hubspot staging models.
+    - Updated `int_rag_hubspot__deal_document` to adjust the method that `hubspot_engagement_*` models are joined by leveraging the `hubspot__engagement` table as the intermediary joining table for the `engagement_contact` and `engagement_company` tables.
```
```suggestion
    - Updated `int_rag_hubspot__deal_document` joins so that `stg_rag_hubspot__engagement` table served as the intermediary joining table for the `stg_rag_hubspot__engagement_contact` and `stg_rag_hubspot__engagement_company` tables.
```

This was a bit confusing to read. I reworded it, let me know your thoughts.

fivetran-avinash commented on CHANGELOG.md:
```diff
@@ -1,3 +1,14 @@
+    - Updated `int_rag_hubspot__deal_document` to retrieve `engagement_type` from the hubspot `engagement` table as opposed to the `engagement_emails` and `engagement_notes` tables. As such, removes their respective references as they are no longer used in this model.
+
+## Bug Fix (`--full-refresh` required when upgrading)
+- Updated the unique key in `rag__unified_document` to include `chunk_index`. Previously, the unique key was a combination of only `document_id`, `platform`, and `source_relation`, which was potentially inaccurate if there were multiple chunks associated with a document.
```
```suggestion
- Updated the `unique_id` in `rag__unified_document` to include `chunk_index`. Previously, the unique key was a combination of only `document_id`, `platform`, and `source_relation`, which was potentially inaccurate if there were multiple chunks associated with a document.
```

fivetran-avinash reviewed (commented):
@fivetran-reneeli Thanks for the PR! A few comments and a question about the seed file before approving.

fivetran-reneeli commented:
Thanks @fivetran-avinash ! Addressed comments. Also, I noticed that I didn't update the hubspot seed data with the new `category` field. Even though it doesn't play a part downstream, I figured to update the seed data and get-columns macro anyways. Additionally, I noticed some missing field descriptions in the rest of the hubspot staging models, so I added those in as well.

fivetran-avinash commented on CHANGELOG.md:
```diff
@@ -1,3 +1,18 @@
+- Updated the `unique_id` in `rag__unified_document` to include `chunk_index`. Previously, the unique key was a combination of only `document_id`, `platform`, and `source_relation`, which was potentially inaccurate if there were multiple chunks associated with a document.
+
+## Under the Hood
+- Updated the *hubspot_x* seed data and *get_hubspot_x_columns* macros with the new `category` field where relevant.
```
```suggestion
- Updated the `hubspot_*` seed data and `get_hubspot_*_columns` macros with the new `category` field where relevant.
```

fivetran-avinash reviewed (commented):
Thanks @fivetran-reneeli for doing this quick review!

Good call out on adding `category` to the relevant seed files. However, I noticed `category` is not being brought into the staging models, and we need to  account for the joins in `int_rag_hubspot__deal_document`, as was discussed in [#10] (I presume it's part of this PR since it's linked to the issue). Sorry for not catching this in the first go-around.
 
 @fivetran-joemarkiewicz  Let's discuss Monday if we want to try and fold these changes back in this sprint or just release as is.

fivetran-avinash commented on integration_tests/seeds/hubspot_engagement.csv:
```diff
@@ -0,0 +1,11 @@
+id,_fivetran_synced,active,activity_type,created_at,last_updated,owner_id,portal_id,timestamp,type
+19732910159,2023-11-02 13:50:33.715000 UTC,,,,,,4703379,,CALL
```
I think we will need to remove these UTC strings for Snowflake Buildkite errors to be fixed.

fivetran-avinash commented on integration_tests/seeds/hubspot_engagement.csv:
```diff
@@ -0,0 +1,11 @@
+id,_fivetran_synced,active,activity_type,created_at,last_updated,owner_id,portal_id,timestamp,type
```
I think we will need to remove UTC strings on the `_fivetran_synced` fields for Snowflake Buildkite errors to be fixed.

fivetran-avinash replied:
This fixed it!

fivetran-avinash reviewed (commented):
Suggestion on how to resolve Snowflake Buildkite errors.

fivetran-avinash commented:
Removed the UTC strings and this fixed the Buildkite errors. With @fivetran-reneeli out, I will merge this onto main and cut as a pre-release.
